File: Python/PID.py
import csv
import time


class PID:
    def __init__(self, P: int = 1, I: int = 0, D: int = 0, debug: bool = True, file: str = "PIDvars.csv") -> None:
        self.P = P
        self.I = I
        self.D = D
        self.debug = debug
        self.iAccumulator = 0
        self.prevError = 0
        if self.debug:
            self.fileOutput = open(file, "w")
            self.writePointer = csv.writer(self.fileOutput)
            self.writePointer.writerow([
                "Equation", "I Accumulator", "Error", "Prev Error",
                "P With Error", "I with I Accumulator", "D with Prev Error",
                "Setpoint", "PIDInput", "Time"
            ])
        self.first = True

    def update(self, target: int, current: int) -> None:
        error = target - current
        self.iAccumulator += error
        if self.first:
            self.iAccumulator = 0
            self.prevError = error
            self.first = False
            self.sTime = time.time()
        output = (self.P * error) + (self.iAccumulator * self.I) + ((error - self.prevError) * self.D)
        if self.debug:
            self.writePointer.writerow([
                output,
                self.iAccumulator,
                error,
                self.prevError,
                self.P * error,
                self.I * self.iAccumulator,
                self.D * (error-self.prevError),
                target,
                current,
                time.time() - self.sTime
            ])
        self.prevError = error

        return output

    def reset(self) -> None:
        self.first = True

    def close(self) -> None:
        if self.debug:
            self.fileOutput.close()

File: Javascript/PID.ts
import fs from "fs";
import os from "os";


/*
    Use import() to load this module from commonjs. 
    https://nodejs.org/api/esm.html#esm_import_expression
*/
export default class PID {
    D: number;
    I: number;
    P: number;
    debug: boolean;
    iAccumulator: number;
    prevError: number;
    first: boolean;
    file: string;
    sTime: number;

    constructor(P: number = 1, I: number = 0, D: number = 0, debug: boolean = true, file: string = "PIDvars.csv") {
        this.P = P;
        this.I = I;
        this.D = D;
        this.debug = debug;
        this.iAccumulator = 0;
        this.prevError = 0;
        this.first = true;
        this.file = file;
        fs.writeFileSync(this.file, `Equation,I Accumulator,Error,Prev Error,P With Error,I with I Accumulator,D with Prev Error,Setpoint,Time${os.EOL}`);
    }

    update(target: number, current: number) {
        let error = target - current;
        this.iAccumulator += error;
        if (this.first) {
            this.iAccumulator = 0;
            this.prevError = error;
            this.first = false;
            this.sTime = new Date().getTime() / 1000;
        }
        let output = (this.P * error) + (this.iAccumulator * this.I) + ((error - this.prevError) * this.D);
        fs.appendFileSync(this.file, [
            output,
            this.iAccumulator,
            error,
            this.prevError,
            this.P * error,
            this.I * this.iAccumulator,
            this.D * (error-this.prevError),
            target,
            current,
            (new Date().getTime() / 1000) - this.sTime
        ].join(",") + os.EOL);
        return output;
    }

    reset() {
        this.first = true;
    }
}
